cpp/ql/src/semmle/code/cpp/models/implementations/StdContainer.qll

```diff
@@ -26,9 +27,17 @@ class StdSequenceContainerConstructor extends Constructor, TaintFunction {
       getDeclaringType().getTemplateArgument(0).(Type).getUnspecifiedType() // i.e. the `T` of this `std::vector<T>`
   }
 
+  /**
+   * Gets the index of a parameter to this function that is an iterator.
+   */
+  int getAnIteratorParameterIndex() { getParameter(result).getType() instanceof Iterator }
+
   override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
     // taint flow from any parameter of the value type to the returned object
-    input.isParameterDeref(getAValueTypeParameterIndex()) and
+    (
+      input.isParameterDeref(getAValueTypeParameterIndex()) or
+      input.isParameter(getAnIteratorParameterIndex())
```

Comment:
This implies that we get taint from the `end` iterator as well, right? Given that the convention is that all `end` iterators return a value that is "1 passed the last entry", is this what we want?

Reply:
In general you can get taint from an `end` iterator, at least if it's a bidirectional iterator, by iterating backwards from it.  There's a [test case](https://github.com/github/codeql/blob/main/cpp/ql/test/library-tests/dataflow/taint-tests/string.cpp#L413) for this (not that I would expect it to occur much in the wild).

In this particular case, I believe the `begin` and `end` iterators are required to point to the same container so it's largely irrelevant whether taint flows from `begin` and `end` or just `begin`.  I suppose this could change if we ever model *part* of a container being tainted and track which iterators point to the tainted bit.  But I *think* at the moment as well as being simpler, the current approach is more robust in that it supports unanticipated constructors (such as might appear in a new version of the standard).